File: android/app/src/main/java/com/golfiq/hud/analytics/AnalyticsController.kt
package com.golfiq.hud.analytics

import android.content.Context
import com.golfiq.hud.model.FeatureFlagConfig
import io.sentry.SentryEvent
import io.sentry.SentryOptions
import io.sentry.android.core.SentryAndroid
import io.sentry.protocol.SentryException
import io.sentry.protocol.User
import java.net.URL
import java.util.Locale
import kotlin.random.Random

class AnalyticsController(
    context: Context,
    baseUrl: URL,
) {
    private val consentStore = AnalyticsConsentStore(context)
    private val crashBridge = CrashTelemetryBridge(baseUrl)
    private val envAnalyticsEnabled = System.getenv("ANALYTICS_ENABLED")?.lowercase(Locale.US) != "false"
    private val sentryDsn = System.getenv("SENTRY_DSN_MOBILE")?.takeIf { it.isNotBlank() }
    private val random = Random(System.currentTimeMillis())
    private val appContext = context.applicationContext

    @Volatile
    private var remoteAnalyticsEnabled: Boolean = true

    @Volatile
    private var remoteCrashEnabled: Boolean = true

    @Volatile
    private var sentryInstalled: Boolean = false

    @Volatile
    private var crashHandlerInstalled: Boolean = false

    private var upstreamHandler: Thread.UncaughtExceptionHandler? = null

    fun update(featureFlags: FeatureFlagConfig) {
        remoteAnalyticsEnabled = featureFlags.analyticsEnabled
        remoteCrashEnabled = featureFlags.crashEnabled
        if (isAnalyticsAllowed()) {
            installSentry()
        } else {
            clearSentry()
        }
        installCrashHandler()
    }

    fun setUserConsent(granted: Boolean) {
        consentStore.setGranted(granted)
        if (!granted) {
            clearSentry()
        } else if (isAnalyticsAllowed()) {
            installSentry()
        }
    }

    fun hasUserConsent(): Boolean = consentStore.isGranted()

    private fun isAnalyticsAllowed(): Boolean {
        return envAnalyticsEnabled && consentStore.isGranted() && remoteAnalyticsEnabled
    }

    private fun isCrashAllowed(): Boolean {
        return isAnalyticsAllowed() && remoteCrashEnabled
    }

    private fun installSentry() {
        if (sentryInstalled || sentryDsn.isNullOrBlank()) {
            return
        }
        runCatching {
            SentryAndroid.init(appContext) { options ->
                configureOptions(options)
            }
            sentryInstalled = true
        }.onFailure {
            sentryInstalled = false
        }
    }

    private fun configureOptions(options: SentryOptions) {
        options.dsn = sentryDsn
        options.isEnableAutoSessionTracking = false
        options.isEnableUserInteractionBreadcrumbs = false
        options.isSendDefaultPii = false
        options.beforeSend = SentryOptions.BeforeSendCallback { event: SentryEvent, _ ->
            if (!isAnalyticsAllowed() || random.nextDouble() > SAMPLE_RATE) {
                return@BeforeSendCallback null
            }
            event.user = event.user?.scrubbed() ?: User()
            event.contexts.remove("geo")
            event.exceptions?.forEach { exception ->
                scrubStack(exception)
            }
            event
        }
    }

    private fun scrubStack(exception: SentryException) {
        val stacktrace = exception.stacktrace ?: return
        val frames = stacktrace.frames ?: return
        if (frames.size > MAX_STACK_FRAMES) {
            stacktrace.frames = frames.takeLast(MAX_STACK_FRAMES)
        }
    }

    private fun clearSentry() {
        if (!sentryInstalled) {
            return
        }
        runCatching {
            io.sentry.Sentry.close()
        }
        sentryInstalled = false
    }

    private fun installCrashHandler() {
        if (crashHandlerInstalled) {
            return
        }
        upstreamHandler = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
            crashBridge.postCrash(throwable, isAnalyticsAllowed(), isCrashAllowed())
            upstreamHandler?.uncaughtException(thread, throwable)
        }
        crashHandlerInstalled = true
    }

    fun shutdown() {
        if (crashHandlerInstalled) {
            Thread.setDefaultUncaughtExceptionHandler(upstreamHandler)
            crashHandlerInstalled = false
        }
        clearSentry()
        crashBridge.shutdown()
    }

    private fun User.scrubbed(): User {
        this.email = null
        this.ipAddress = null
        return this
    }

    private companion object {
        private const val SAMPLE_RATE = 0.2
        private const val MAX_STACK_FRAMES = 50
    }
}

File: ios/Services/AnalyticsController.swift
import Foundation

#if canImport(Sentry)
import Sentry
#endif

final class AnalyticsController {
    private let consentStore: AnalyticsConsentStore
    private let crashBridge: CrashTelemetryBridge
    private let envAnalyticsEnabled: Bool
    private let sentryDSN: String?
    private var remoteAnalyticsEnabled: Bool = true
    private var remoteCrashEnabled: Bool = true
    private var sentryInstalled = false
    private var uncaughtHandlerInstalled = false
    private var previousExceptionHandler: (@convention(c) (NSException) -> Void)?
    private let sampleRate: Double = 0.2

    init(baseURL: URL, consentStore: AnalyticsConsentStore = AnalyticsConsentStore()) {
        self.consentStore = consentStore
        self.crashBridge = CrashTelemetryBridge(baseURL: baseURL)
        let env = ProcessInfo.processInfo.environment
        self.envAnalyticsEnabled = env["ANALYTICS_ENABLED"]?.lowercased() != "false"
        self.sentryDSN = env["SENTRY_DSN_MOBILE"].flatMap { $0.isEmpty ? nil : $0 }
    }

    func update(flags: FeatureFlagConfig) {
        remoteAnalyticsEnabled = flags.analyticsEnabled
        remoteCrashEnabled = flags.crashEnabled
        if isAnalyticsAllowed {
            installSentry()
        } else {
            shutdownSentry()
        }
        installCrashHandler()
    }

    func setConsent(granted: Bool) {
        consentStore.setGranted(granted)
        if granted {
            if isAnalyticsAllowed {
                installSentry()
            }
        } else {
            shutdownSentry()
        }
    }

    var hasConsent: Bool { consentStore.isGranted }

    private var isAnalyticsAllowed: Bool {
        envAnalyticsEnabled && consentStore.isGranted && remoteAnalyticsEnabled
    }

    private func installCrashHandler() {
        guard !uncaughtHandlerInstalled else { return }
        previousExceptionHandler = NSGetUncaughtExceptionHandler()
        AnalyticsCrashProxy.controller = self
        NSSetUncaughtExceptionHandler { exception in
            AnalyticsCrashProxy.controller?.handle(exception: exception)
        }
        uncaughtHandlerInstalled = true
    }

    private func handle(exception: NSException) {
        crashBridge.postCrash(exception, analyticsEnabled: isAnalyticsAllowed, crashEnabled: remoteCrashEnabled)
        previousExceptionHandler?(exception)
    }

    #if canImport(Sentry)
    private func installSentry() {
        guard !sentryInstalled, let dsn = sentryDSN else { return }
        SentrySDK.start { [weak self] options in
            guard let self else { return }
            options.dsn = dsn
            options.enableAutoSessionTracking = false
            options.sendDefaultPii = false
            options.beforeSend = { event in
                guard self.isAnalyticsAllowed else { return nil }
                if Double.random(in: 0 ..< 1) > self.sampleRate { return nil }
                event.user?.email = nil
                event.user?.ipAddress = nil
                if var contexts = event.context, contexts["geo"] != nil {
                    contexts["geo"] = nil
                    event.context = contexts
                }
                if var exceptions = event.exceptions, !exceptions.isEmpty {
                    for index in exceptions.indices {
                        if var stacktrace = exceptions[index].stacktrace,
                           stacktrace.frames.count > self.maxFrames {
                            stacktrace.frames = Array(stacktrace.frames.suffix(self.maxFrames))
                            exceptions[index].stacktrace = stacktrace
                        }
                    }
                    event.exceptions = exceptions
                }
                return event
            }
        }
        sentryInstalled = true
    }
    #else
    private func installSentry() {
        // No-op when Sentry SDK is not linked.
    }
    #endif

    func shutdown() {
        shutdownSentry()
        if uncaughtHandlerInstalled {
            NSSetUncaughtExceptionHandler(previousExceptionHandler)
            AnalyticsCrashProxy.controller = nil
            uncaughtHandlerInstalled = false
        }
    }

    private func shutdownSentry() {
        #if canImport(Sentry)
        if sentryInstalled {
            SentrySDK.close()
            sentryInstalled = false
        }
        #endif
    }
}

private let maxFramesDefault = 50

extension AnalyticsController {
    fileprivate var maxFrames: Int { maxFramesDefault }
}

private final class AnalyticsCrashProxy {
    static weak var controller: AnalyticsController?
}

private extension CrashTelemetryBridge {
    func postCrash(_ exception: NSException, analyticsEnabled: Bool, crashEnabled: Bool) {
        let error = NSError(domain: exception.name.rawValue, code: 0, userInfo: [NSLocalizedDescriptionKey: exception.reason ?? exception.name.rawValue])
        postCrash(
            error,
            analyticsEnabled: analyticsEnabled,
            crashEnabled: crashEnabled,
            stackSymbols: exception.callStackSymbols
        )
    }
}
